apps/nectar/web/calculators/shipping_calculator.ex

```diff
@@ -1,25 +1,116 @@
 defmodule Nectar.ShippingCalculator do
+  use GenServer
+
+  alias __MODULE__
+  alias Nectar.Repo
+  alias Nectar.ShippingMethod
   alias Nectar.Order
   alias Nectar.Repo
 
+  # hold the state in this struct
+  defstruct order: nil, result: [], timer: nil, caller: nil, pending: [], shipping_methods: []
+
+  @shipping_calculation_timeout 5000 # ms
+
   # generate all possible shippings
   def calculate_applicable_shippings(%Order{} = order) do
-    # replace with a distributed worker/async approach
-    # collect all available results avoid awaiting for all
-    # see: http://theerlangelist.com/article/beyond_taskasync for reference
-    Enum.map(Repo.all(Nectar.ShippingMethod.enabled_shipping_methods), fn (shipping_method) -> calculate_shipping_cost(shipping_method, order) end)
+    available_shipping_methods = Repo.all ShippingMethod.enabled_shipping_methods
+    {:ok, server} = ShippingCalculator.start(self(), available_shipping_methods, order)
+    GenServer.cast(server, {:calculate})
+    applicable_shippings = receive do
+      {:ok, results} -> results
+    end
+    applicable_shippings
   end
 
-  def calculate_shipping_cost(shipping_method, order) do
+  def calculate_shipping_cost(%ShippingMethod{} = shipping_method, order) do
     # launch the shipping calculator here.
-    Map.from_struct(%Nectar.ShippingMethod{shipping_method|shipping_cost: shipping_cost(shipping_method, order)})
-    |> Map.drop([:__meta__, :shippings])
+    {status, calculated_shipping_cost} = shipping_cost(shipping_method, order)
+    cost = Map.from_struct(%ShippingMethod{shipping_method|shipping_cost: calculated_shipping_cost})
+    |> Map.drop([:__meta__, :shippings, :inserted_at, :updated_at])
+    {status, cost}
+  end
+
+  defp shipping_calculator_module(method_name) do
+    Application.get_env(:shipping_calculators, String.to_atom(method_name))
+  end
+
+  def shipping_cost(%ShippingMethod{name: name}, order) do
+    shipping_calculator_module(name).calculate_shipping(order)
+  end
+
+  #### Genserver Methods Start Here ####
+
+  def start(caller, shipping_methods, order) do
+    state = %ShippingCalculator{caller: caller, shipping_methods: shipping_methods, order: order}
+    GenServer.start(__MODULE__, state, [])
+  end
+
+  def handle_cast({:calculate}, state) do
+    current = self()
+    proc_list = Enum.map(state.shipping_methods, fn(method) ->
+      spawn_monitor(fn ->
+        send(current, Tuple.append(ShippingCalculator.calculate_shipping_cost(method, state.order), self()))
+      end)
+    end)
+    timer = Process.send_after(current, {:timeout}, @shipping_calculation_timeout)
+    {:noreply, %ShippingCalculator{state|timer: timer, pending: proc_list}}
+  end
+
+  def handle_info({:ok, results, pid}, state) do
```

Comment:
**convention suggestion** What do you think of not having line-spaces between functions who are part of pattern match like *handle_info* here ??